Description: Drop zunclient support as zun is not packaged in Ubuntu.
  This is an alternative to including python-zunclient in main.
  stevedore used to treat client dependencies as warnings if
  they weren't installed but that has since become a hard failure.
  For more details see the bug linked.
Ubuntu-Bug: https://bugs.launchpad.net/cloud-archive/+bug/1893935
Forwarded: no

--- a/requirements.txt
+++ b/requirements.txt
@@ -59,7 +59,6 @@
 python-troveclient>=2.2.0 # Apache-2.0
 python-vitrageclient>=2.7.0 # Apache-2.0
 python-zaqarclient>=1.3.0 # Apache-2.0
-python-zunclient>=3.4.0 # Apache-2.0
 pytz>=2013.6 # MIT
 PyYAML>=5.1 # MIT
 requests>=2.23.0 # Apache-2.0
--- a/heat/tests/openstack/zun/test_container.py
+++ /dev/null
@@ -1,549 +0,0 @@
-#
-#    Licensed under the Apache License, Version 2.0 (the "License"); you may
-#    not use this file except in compliance with the License. You may obtain
-#    a copy of the License at
-#
-#         http://www.apache.org/licenses/LICENSE-2.0
-#
-#    Unless required by applicable law or agreed to in writing, software
-#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
-#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
-#    License for the specific language governing permissions and limitations
-#    under the License.
-
-import copy
-from unittest import mock
-
-from oslo_config import cfg
-from zunclient import exceptions as zc_exc
-
-from heat.common import exception
-from heat.common import template_format
-from heat.engine.clients.os import neutron
-from heat.engine.clients.os import zun
-from heat.engine.resources.openstack.zun import container
-from heat.engine import scheduler
-from heat.engine import template
-from heat.tests import common
-from heat.tests import utils
-
-zun_template = '''
-heat_template_version: 2017-09-01
-
-resources:
-  test_container:
-    type: OS::Zun::Container
-    properties:
-      name: test_container
-      image: "cirros:latest"
-      command: sleep 10000
-      cpu: 0.1
-      memory: 100
-      environment:
-        myenv: foo
-      workdir: /testdir
-      labels:
-        mylabel: bar
-      image_pull_policy: always
-      restart_policy: on-failure:2
-      interactive: false
-      tty: false
-      image_driver: docker
-      hints:
-        hintkey: hintval
-      hostname: myhost
-      security_groups:
-        - my_seg
-      mounts:
-        - volume_size: 1
-          mount_path: /data
-        - volume_id: 6ec29ba3-bf2c-4276-a88e-3670ea5abc80
-          mount_path: /data2
-      networks:
-        - network: mynet
-          fixed_ip: 10.0.0.4
-        - network: mynet2
-          fixed_ip: fe80::3
-        - port: myport
-'''
-
-zun_template_minimum = '''
-heat_template_version: 2017-09-01
-
-resources:
-  test_container:
-    type: OS::Zun::Container
-    properties:
-      name: test_container
-      image: "cirros:latest"
-'''
-
-
-def create_fake_iface(port=None, net=None, mac=None, ip=None, subnet=None):
-    class fake_interface(object):
-        def __init__(self, port_id, net_id, mac_addr, fixed_ip, subnet_id):
-            self.port_id = port_id
-            self.net_id = net_id
-            self.mac_addr = mac_addr
-            self.fixed_ips = [{'ip_address': fixed_ip, 'subnet_id': subnet_id}]
-
-    return fake_interface(port, net, mac, ip, subnet)
-
-
-class ZunContainerTest(common.HeatTestCase):
-
-    def setUp(self):
-        super(ZunContainerTest, self).setUp()
-
-        self.resource_id = '12345'
-        self.fake_name = 'test_container'
-        self.fake_image = 'cirros:latest'
-        self.fake_command = 'sleep 10000'
-        self.fake_cpu = 0.1
-        self.fake_memory = 100
-        self.fake_env = {'myenv': 'foo'}
-        self.fake_workdir = '/testdir'
-        self.fake_labels = {'mylabel': 'bar'}
-        self.fake_image_policy = 'always'
-        self.fake_restart_policy = {'MaximumRetryCount': '2',
-                                    'Name': 'on-failure'}
-        self.fake_interactive = False
-        self.fake_tty = False
-        self.fake_image_driver = 'docker'
-        self.fake_hints = {'hintkey': 'hintval'}
-        self.fake_hostname = 'myhost'
-        self.fake_security_groups = ['my_seg']
-        self.fake_mounts = [
-            {'volume_id': None, 'volume_size': 1, 'mount_path': '/data'},
-            {'volume_id': '6ec29ba3-bf2c-4276-a88e-3670ea5abc80',
-             'volume_size': None, 'mount_path': '/data2'}]
-        self.fake_mounts_args = [
-            {'size': 1, 'destination': '/data'},
-            {'source': '6ec29ba3-bf2c-4276-a88e-3670ea5abc80',
-             'destination': '/data2'}]
-        self.fake_networks = [
-            {'network': 'mynet', 'port': None, 'fixed_ip': '10.0.0.4'},
-            {'network': 'mynet2', 'port': None, 'fixed_ip': 'fe80::3'},
-            {'network': None, 'port': 'myport', 'fixed_ip': None}]
-        self.fake_networks_args = [
-            {'network': 'mynet', 'v4-fixed-ip': '10.0.0.4'},
-            {'network': 'mynet2', 'v6-fixed-ip': 'fe80::3'},
-            {'port': 'myport'}]
-
-        self.fake_network_id = '9c11d847-99ce-4a83-82da-9827362a68e8'
-        self.fake_network_name = 'private'
-        self.fake_networks_attr = {
-            'networks': [
-                {
-                    'id': self.fake_network_id,
-                    'name': self.fake_network_name,
-                }
-            ]
-        }
-        self.fake_address = {
-            'version': 4,
-            'addr': '10.0.0.12',
-            'port': 'ab5c12d8-f414-48a3-b765-8ce34a6714d2'
-        }
-        self.fake_addresses = {
-            self.fake_network_id: [self.fake_address]
-        }
-        self.fake_extended_addresses = {
-            self.fake_network_id: [self.fake_address],
-            self.fake_network_name: [self.fake_address],
-        }
-
-        t = template_format.parse(zun_template)
-        self.stack = utils.parse_stack(t)
-        resource_defns = self.stack.t.resource_definitions(self.stack)
-        self.rsrc_defn = resource_defns[self.fake_name]
-        self.client = mock.Mock()
-        self.patchobject(container.Container, 'client',
-                         return_value=self.client)
-        self.neutron_client = mock.Mock()
-        self.patchobject(container.Container, 'neutron',
-                         return_value=self.neutron_client)
-        self.stub_VolumeConstraint_validate()
-        self.mock_update = self.patchobject(zun.ZunClientPlugin,
-                                            'update_container')
-        self.stub_PortConstraint_validate()
-        self.mock_find = self.patchobject(
-            neutron.NeutronClientPlugin,
-            'find_resourceid_by_name_or_id',
-            side_effect=lambda x, y: y)
-        self.mock_attach = self.patchobject(zun.ZunClientPlugin,
-                                            'network_attach')
-        self.mock_detach = self.patchobject(zun.ZunClientPlugin,
-                                            'network_detach')
-        self.mock_attach_check = self.patchobject(zun.ZunClientPlugin,
-                                                  'check_network_attach',
-                                                  return_value=True)
-        self.mock_detach_check = self.patchobject(zun.ZunClientPlugin,
-                                                  'check_network_detach',
-                                                  return_value=True)
-
-    def _mock_get_client(self):
-        value = mock.MagicMock()
-        value.name = self.fake_name
-        value.image = self.fake_image
-        value.command = self.fake_command
-        value.cpu = self.fake_cpu
-        value.memory = self.fake_memory
-        value.environment = self.fake_env
-        value.workdir = self.fake_workdir
-        value.labels = self.fake_labels
-        value.image_pull_policy = self.fake_image_policy
-        value.restart_policy = self.fake_restart_policy
-        value.interactive = self.fake_interactive
-        value.tty = self.fake_tty
-        value.image_driver = self.fake_image_driver
-        value.hints = self.fake_hints
-        value.hostname = self.fake_hostname
-        value.security_groups = self.fake_security_groups
-        value.addresses = self.fake_addresses
-        value.to_dict.return_value = value.__dict__
-
-        self.client.containers.get.return_value = value
-
-    def _create_resource(self, name, snippet, stack, status='Running'):
-        value = mock.MagicMock(uuid=self.resource_id)
-        self.client.containers.run.return_value = value
-        get_rv = mock.MagicMock(status=status)
-        self.client.containers.get.return_value = get_rv
-        c = container.Container(name, snippet, stack)
-        return c
-
-    def test_create(self):
-        c = self._create_resource('container', self.rsrc_defn,
-                                  self.stack)
-        # validate the properties
-        self.assertEqual(
-            self.fake_name,
-            c.properties.get(container.Container.NAME))
-        self.assertEqual(
-            self.fake_image,
-            c.properties.get(container.Container.IMAGE))
-        self.assertEqual(
-            self.fake_command,
-            c.properties.get(container.Container.COMMAND))
-        self.assertEqual(
-            self.fake_cpu,
-            c.properties.get(container.Container.CPU))
-        self.assertEqual(
-            self.fake_memory,
-            c.properties.get(container.Container.MEMORY))
-        self.assertEqual(
-            self.fake_env,
-            c.properties.get(container.Container.ENVIRONMENT))
-        self.assertEqual(
-            self.fake_workdir,
-            c.properties.get(container.Container.WORKDIR))
-        self.assertEqual(
-            self.fake_labels,
-            c.properties.get(container.Container.LABELS))
-        self.assertEqual(
-            self.fake_image_policy,
-            c.properties.get(container.Container.IMAGE_PULL_POLICY))
-        self.assertEqual(
-            'on-failure:2',
-            c.properties.get(container.Container.RESTART_POLICY))
-        self.assertEqual(
-            self.fake_interactive,
-            c.properties.get(container.Container.INTERACTIVE))
-        self.assertEqual(
-            self.fake_tty,
-            c.properties.get(container.Container.TTY))
-        self.assertEqual(
-            self.fake_image_driver,
-            c.properties.get(container.Container.IMAGE_DRIVER))
-        self.assertEqual(
-            self.fake_hints,
-            c.properties.get(container.Container.HINTS))
-        self.assertEqual(
-            self.fake_hostname,
-            c.properties.get(container.Container.HOSTNAME))
-        self.assertEqual(
-            self.fake_security_groups,
-            c.properties.get(container.Container.SECURITY_GROUPS))
-        self.assertEqual(
-            self.fake_mounts,
-            c.properties.get(container.Container.MOUNTS))
-        self.assertEqual(
-            self.fake_networks,
-            c.properties.get(container.Container.NETWORKS))
-
-        scheduler.TaskRunner(c.create)()
-        self.assertEqual(self.resource_id, c.resource_id)
-        self.assertEqual((c.CREATE, c.COMPLETE), c.state)
-        self.assertEqual('containers', c.entity)
-        self.client.containers.run.assert_called_once_with(
-            name=self.fake_name,
-            image=self.fake_image,
-            command=self.fake_command.split(),
-            cpu=self.fake_cpu,
-            memory=self.fake_memory,
-            environment=self.fake_env,
-            workdir=self.fake_workdir,
-            labels=self.fake_labels,
-            image_pull_policy=self.fake_image_policy,
-            restart_policy=self.fake_restart_policy,
-            interactive=self.fake_interactive,
-            tty=self.fake_tty,
-            image_driver=self.fake_image_driver,
-            hints=self.fake_hints,
-            hostname=self.fake_hostname,
-            security_groups=self.fake_security_groups,
-            mounts=self.fake_mounts_args,
-            nets=self.fake_networks_args,
-        )
-
-    def test_container_create_failed(self):
-        cfg.CONF.set_override('action_retry_limit', 0)
-        c = self._create_resource('container', self.rsrc_defn, self.stack,
-                                  status='Error')
-        exc = self.assertRaises(
-            exception.ResourceFailure,
-            scheduler.TaskRunner(c.create))
-        self.assertEqual((c.CREATE, c.FAILED), c.state)
-        self.assertIn("Error in creating container ", str(exc))
-
-    def test_container_create_unknown_status(self):
-        c = self._create_resource('container', self.rsrc_defn, self.stack,
-                                  status='FOO')
-        exc = self.assertRaises(
-            exception.ResourceFailure,
-            scheduler.TaskRunner(c.create))
-        self.assertEqual((c.CREATE, c.FAILED), c.state)
-        self.assertIn("Unknown status Container", str(exc))
-
-    def test_container_update(self):
-        c = self._create_resource('container', self.rsrc_defn, self.stack)
-        scheduler.TaskRunner(c.create)()
-        t = template_format.parse(zun_template)
-        new_t = copy.deepcopy(t)
-        new_t['resources'][self.fake_name]['properties']['name'] = \
-            'fake-container'
-        new_t['resources'][self.fake_name]['properties']['cpu'] = 10
-        new_t['resources'][self.fake_name]['properties']['memory'] = 10
-        rsrc_defns = template.Template(new_t).resource_definitions(self.stack)
-        new_c = rsrc_defns[self.fake_name]
-        scheduler.TaskRunner(c.update, new_c)()
-        self.mock_update.assert_called_once_with(
-            self.resource_id,
-            cpu=10, memory=10, name='fake-container')
-        self.assertEqual((c.UPDATE, c.COMPLETE), c.state)
-
-    def _test_container_update_None_networks(self, new_networks):
-        t = template_format.parse(zun_template_minimum)
-        stack = utils.parse_stack(t)
-        resource_defns = stack.t.resource_definitions(stack)
-        rsrc_defn = resource_defns[self.fake_name]
-        c = self._create_resource('container', rsrc_defn, stack)
-        scheduler.TaskRunner(c.create)()
-
-        new_t = copy.deepcopy(t)
-        new_t['resources'][self.fake_name]['properties']['networks'] = \
-            new_networks
-        rsrc_defns = template.Template(new_t).resource_definitions(stack)
-        new_c = rsrc_defns[self.fake_name]
-        iface = create_fake_iface(
-            port='aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa',
-            net='450abbc9-9b6d-4d6f-8c3a-c47ac34100ef',
-            ip='1.2.3.4')
-        self.client.containers.network_list.return_value = [iface]
-        scheduler.TaskRunner(c.update, new_c)()
-        self.assertEqual((c.UPDATE, c.COMPLETE), c.state)
-        self.client.containers.network_list.assert_called_once_with(
-            self.resource_id)
-
-    def test_container_update_None_networks_with_port(self):
-        new_networks = [{'port': '2a60cbaa-3d33-4af6-a9ce-83594ac546fc'}]
-        self._test_container_update_None_networks(new_networks)
-        self.assertEqual(1, self.mock_attach.call_count)
-        self.assertEqual(1, self.mock_detach.call_count)
-        self.assertEqual(1, self.mock_attach_check.call_count)
-        self.assertEqual(1, self.mock_detach_check.call_count)
-
-    def test_container_update_None_networks_with_network_id(self):
-        new_networks = [{'network': 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa',
-                         'fixed_ip': '1.2.3.4'}]
-        self._test_container_update_None_networks(new_networks)
-        self.assertEqual(1, self.mock_attach.call_count)
-        self.assertEqual(1, self.mock_detach.call_count)
-        self.assertEqual(1, self.mock_attach_check.call_count)
-        self.assertEqual(1, self.mock_detach_check.call_count)
-
-    def test_container_update_None_networks_with_complex_parameters(self):
-        new_networks = [{'network': 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa',
-                         'fixed_ip': '1.2.3.4',
-                         'port': '2a60cbaa-3d33-4af6-a9ce-83594ac546fc'}]
-        self._test_container_update_None_networks(new_networks)
-        self.assertEqual(1, self.mock_attach.call_count)
-        self.assertEqual(1, self.mock_detach.call_count)
-        self.assertEqual(1, self.mock_attach_check.call_count)
-        self.assertEqual(1, self.mock_detach_check.call_count)
-
-    def test_server_update_empty_networks_to_None(self):
-        new_networks = None
-        self._test_container_update_None_networks(new_networks)
-        self.assertEqual(0, self.mock_attach.call_count)
-        self.assertEqual(0, self.mock_detach.call_count)
-        self.assertEqual(0, self.mock_attach_check.call_count)
-        self.assertEqual(0, self.mock_detach_check.call_count)
-
-    def _test_container_update_networks(self, new_networks):
-        c = self._create_resource('container', self.rsrc_defn, self.stack)
-        scheduler.TaskRunner(c.create)()
-        t = template_format.parse(zun_template)
-        new_t = copy.deepcopy(t)
-        new_t['resources'][self.fake_name]['properties']['networks'] = \
-            new_networks
-        rsrc_defns = template.Template(new_t).resource_definitions(self.stack)
-        new_c = rsrc_defns[self.fake_name]
-        sec_uuids = ['86c0f8ae-23a8-464f-8603-c54113ef5467']
-        self.patchobject(neutron.NeutronClientPlugin,
-                         'get_secgroup_uuids', return_value=sec_uuids)
-        ifaces = [
-            create_fake_iface(port='95e25541-d26a-478d-8f36-ae1c8f6b74dc',
-                              net='mynet',
-                              ip='10.0.0.4'),
-            create_fake_iface(port='450abbc9-9b6d-4d6f-8c3a-c47ac34100ef',
-                              net='mynet2',
-                              ip='fe80::3'),
-            create_fake_iface(port='myport',
-                              net='aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa',
-                              ip='21.22.23.24')]
-        self.client.containers.network_list.return_value = ifaces
-        scheduler.TaskRunner(c.update, new_c)()
-        self.assertEqual((c.UPDATE, c.COMPLETE), c.state)
-        self.client.containers.network_list.assert_called_once_with(
-            self.resource_id)
-
-    def test_container_update_networks_with_complex_parameters(self):
-        new_networks = [
-            {'network': 'mynet',
-             'fixed_ip': '10.0.0.4'},
-            {'port': '2a60cbaa-3d33-4af6-a9ce-83594ac546fc'}]
-        self._test_container_update_networks(new_networks)
-        self.assertEqual(2, self.mock_detach.call_count)
-        self.assertEqual(1, self.mock_attach.call_count)
-        self.assertEqual(2, self.mock_detach_check.call_count)
-        self.assertEqual(1, self.mock_attach_check.call_count)
-
-    def test_container_update_networks_with_None(self):
-        new_networks = None
-        self._test_container_update_networks(new_networks)
-        self.assertEqual(3, self.mock_detach.call_count)
-        self.assertEqual(1, self.mock_attach.call_count)
-        self.assertEqual(3, self.mock_detach_check.call_count)
-        self.assertEqual(1, self.mock_attach_check.call_count)
-
-    def test_container_update_old_networks_to_empty_list(self):
-        new_networks = []
-        self._test_container_update_networks(new_networks)
-        self.assertEqual(3, self.mock_detach.call_count)
-        self.assertEqual(1, self.mock_attach.call_count)
-        self.assertEqual(3, self.mock_detach_check.call_count)
-        self.assertEqual(1, self.mock_attach_check.call_count)
-
-    def test_container_update_remove_network_non_empty(self):
-        new_networks = [
-            {'network': 'mynet',
-             'fixed_ip': '10.0.0.4'},
-            {'port': 'myport'}]
-        self._test_container_update_networks(new_networks)
-        self.assertEqual(1, self.mock_detach.call_count)
-        self.assertEqual(0, self.mock_attach.call_count)
-        self.assertEqual(1, self.mock_detach_check.call_count)
-        self.assertEqual(0, self.mock_attach_check.call_count)
-
-    def test_container_delete(self):
-        c = self._create_resource('container', self.rsrc_defn, self.stack)
-        scheduler.TaskRunner(c.create)()
-        self.patchobject(self.client.containers, 'get',
-                         side_effect=[c, zc_exc.NotFound('Not Found')])
-        scheduler.TaskRunner(c.delete)()
-        self.assertEqual((c.DELETE, c.COMPLETE), c.state)
-        self.client.containers.delete.assert_called_once_with(
-            c.resource_id, stop=True)
-
-    def test_container_delete_not_found(self):
-        c = self._create_resource('container', self.rsrc_defn, self.stack)
-        scheduler.TaskRunner(c.create)()
-        c.client_plugin = mock.MagicMock()
-        self.client.containers.delete.side_effect = Exception('Not Found')
-        scheduler.TaskRunner(c.delete)()
-        self.assertEqual((c.DELETE, c.COMPLETE), c.state)
-        self.client.containers.delete.assert_called_once_with(
-            c.resource_id, stop=True)
-        mock_ignore_not_found = c.client_plugin.return_value.ignore_not_found
-        self.assertEqual(1, mock_ignore_not_found.call_count)
-
-    def test_container_get_live_state(self):
-        c = self._create_resource('container', self.rsrc_defn, self.stack)
-        scheduler.TaskRunner(c.create)()
-        self._mock_get_client()
-        reality = c.get_live_state(c.properties)
-        self.assertEqual(
-            {
-                container.Container.NAME: self.fake_name,
-                container.Container.CPU: self.fake_cpu,
-                container.Container.MEMORY: self.fake_memory,
-            }, reality)
-
-    def test_resolve_attributes(self):
-        self.neutron_client.list_networks.return_value = \
-            self.fake_networks_attr
-        c = self._create_resource('container', self.rsrc_defn, self.stack)
-        scheduler.TaskRunner(c.create)()
-        self._mock_get_client()
-        self.assertEqual(
-            self.fake_name,
-            c._resolve_attribute(container.Container.NAME))
-        self.assertEqual(
-            self.fake_extended_addresses,
-            c._resolve_attribute(container.Container.ADDRESSES))
-
-    def test_resolve_attributes_duplicate_net_name(self):
-        self.neutron_client.list_networks.return_value = {
-            'networks': [
-                {'id': 'fake_net_id', 'name': 'test'},
-                {'id': 'fake_net_id2', 'name': 'test'},
-            ]
-        }
-        self.fake_addresses = {
-            'fake_net_id': [{'addr': '10.0.0.12'}],
-            'fake_net_id2': [{'addr': '10.100.0.12'}],
-        }
-        self.fake_extended_addresses = {
-            'fake_net_id': [{'addr': '10.0.0.12'}],
-            'fake_net_id2': [{'addr': '10.100.0.12'}],
-            'test': [{'addr': '10.0.0.12'}, {'addr': '10.100.0.12'}],
-        }
-        c = self._create_resource('container', self.rsrc_defn, self.stack)
-        scheduler.TaskRunner(c.create)()
-        self._mock_get_client()
-        self._assert_addresses(
-            self.fake_extended_addresses,
-            c._resolve_attribute(container.Container.ADDRESSES))
-
-    def _assert_addresses(self, expected, actual):
-        matched = True
-        if len(expected) != len(actual):
-            matched = False
-        for key in expected:
-            if key not in actual:
-                matched = False
-                break
-            list1 = expected[key]
-            list1 = sorted(list1, key=lambda x: sorted(x.values()))
-            list2 = actual[key]
-            list2 = sorted(list2, key=lambda x: sorted(x.values()))
-            if list1 != list2:
-                matched = False
-                break
-
-        if not matched:
-            raise AssertionError(
-                'Addresses is unmatched:\n reference = ' + str(expected) +
-                '\nactual = ' + str(actual))
--- a/heat/tests/clients/test_zun_client.py
+++ /dev/null
@@ -1,42 +0,0 @@
-#    Licensed under the Apache License, Version 2.0 (the "License"); you may
-#    not use this file except in compliance with the License. You may obtain
-#    a copy of the License at
-#
-#         http://www.apache.org/licenses/LICENSE-2.0
-#
-#    Unless required by applicable law or agreed to in writing, software
-#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
-#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
-#    License for the specific language governing permissions and limitations
-#    under the License.
-
-from unittest import mock
-
-from heat.tests import common
-from heat.tests import utils
-
-
-class ZunClientPluginTest(common.HeatTestCase):
-
-    def setUp(self):
-        super(ZunClientPluginTest, self).setUp()
-        self.client = mock.Mock()
-        context = utils.dummy_context()
-        self.plugin = context.clients.client_plugin('zun')
-        self.plugin.client = lambda **kw: self.client
-        self.resource_id = '123456'
-
-    def test_create(self):
-        context = utils.dummy_context()
-        plugin = context.clients.client_plugin('zun')
-        client = plugin.client()
-        self.assertEqual('http://server.test:5000/v3',
-                         client.containers.api.session.auth.endpoint)
-        self.assertEqual('1.12',
-                         client.api_version.get_string())
-
-    def test_container_update(self):
-        prop_diff = {'cpu': 10, 'memory': 10, 'name': 'fake-container'}
-        self.plugin.update_container(self.resource_id, **prop_diff)
-        self.client.containers.update.assert_called_once_with(
-            self.resource_id, cpu=10, memory=10, name='fake-container')
--- a/heat/engine/clients/os/zun.py
+++ /dev/null
@@ -1,113 +0,0 @@
-#
-#    Licensed under the Apache License, Version 2.0 (the "License"); you may
-#    not use this file except in compliance with the License. You may obtain
-#    a copy of the License at
-#
-#         http://www.apache.org/licenses/LICENSE-2.0
-#
-#    Unless required by applicable law or agreed to in writing, software
-#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
-#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
-#    License for the specific language governing permissions and limitations
-#    under the License.
-
-from oslo_config import cfg
-import tenacity
-from zunclient import client as zun_client
-from zunclient import exceptions as zc_exc
-
-from heat.engine.clients import client_plugin
-
-CLIENT_NAME = 'zun'
-
-
-class ZunClientPlugin(client_plugin.ClientPlugin):
-
-    service_types = [CONTAINER] = ['container']
-
-    default_version = '1.12'
-
-    supported_versions = [
-        V1_12, V1_18, V1_36,
-    ] = [
-        '1.12', '1.18', '1.36',
-    ]
-
-    def _create(self, version=None):
-        if not version:
-            version = self.default_version
-
-        interface = self._get_client_option(CLIENT_NAME, 'endpoint_type')
-        args = {
-            'interface': interface,
-            'service_type': self.CONTAINER,
-            'session': self.context.keystone_session,
-            'region_name': self._get_region_name()
-        }
-
-        client = zun_client.Client(version, **args)
-        return client
-
-    def update_container(self, container_id, **prop_diff):
-        if prop_diff:
-            self.client(version=self.V1_18).containers.update(
-                container_id, **prop_diff)
-
-    def network_detach(self, container_id, port_id):
-        with self.ignore_not_found:
-            self.client(version=self.V1_18).containers.network_detach(
-                container_id, port=port_id)
-            return True
-
-    def network_attach(self, container_id, port_id=None, net_id=None, fip=None,
-                       security_groups=None):
-        with self.ignore_not_found:
-            kwargs = {}
-            if port_id:
-                kwargs['port'] = port_id
-            if net_id:
-                kwargs['network'] = net_id
-            if fip:
-                kwargs['fixed_ip'] = fip
-            self.client(version=self.V1_18).containers.network_attach(
-                container_id, **kwargs)
-            return True
-
-    @tenacity.retry(
-        stop=tenacity.stop_after_attempt(
-            cfg.CONF.max_interface_check_attempts),
-        wait=tenacity.wait_exponential(multiplier=0.5, max=12.0),
-        retry=tenacity.retry_if_result(client_plugin.retry_if_result_is_false))
-    def check_network_detach(self, container_id, port_id):
-        with self.ignore_not_found:
-            interfaces = self.client(
-                version=self.V1_18).containers.network_list(container_id)
-            for iface in interfaces:
-                if iface.port_id == port_id:
-                    return False
-        return True
-
-    @tenacity.retry(
-        stop=tenacity.stop_after_attempt(
-            cfg.CONF.max_interface_check_attempts),
-        wait=tenacity.wait_exponential(multiplier=0.5, max=12.0),
-        retry=tenacity.retry_if_result(client_plugin.retry_if_result_is_false))
-    def check_network_attach(self, container_id, port_id):
-        if not port_id:
-            return True
-
-        interfaces = self.client(version=self.V1_18).containers.network_list(
-            container_id)
-        for iface in interfaces:
-            if iface.port_id == port_id:
-                return True
-        return False
-
-    def is_not_found(self, ex):
-        return isinstance(ex, zc_exc.NotFound)
-
-    def is_over_limit(self, ex):
-        return isinstance(ex, zc_exc.RequestEntityTooLarge)
-
-    def is_conflict(self, ex):
-        return isinstance(ex, zc_exc.Conflict)
--- a/heat/engine/resources/openstack/zun/container.py
+++ /dev/null
@@ -1,516 +0,0 @@
-#
-#    Licensed under the Apache License, Version 2.0 (the "License"); you may
-#    not use this file except in compliance with the License. You may obtain
-#    a copy of the License at
-#
-#         http://www.apache.org/licenses/LICENSE-2.0
-#
-#    Unless required by applicable law or agreed to in writing, software
-#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
-#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
-#    License for the specific language governing permissions and limitations
-#    under the License.
-
-import copy
-import shlex
-
-from heat.common import exception
-from heat.common.i18n import _
-from heat.engine import attributes
-from heat.engine.clients import progress
-from heat.engine import constraints
-from heat.engine import properties
-from heat.engine import resource
-from heat.engine.resources.openstack.nova import server_network_mixin
-from heat.engine import support
-from heat.engine import translation
-
-
-class Container(resource.Resource,
-                server_network_mixin.ServerNetworkMixin):
-    """A resource that creates a Zun Container.
-
-    This resource creates a Zun container.
-    """
-
-    support_status = support.SupportStatus(version='9.0.0')
-
-    PROPERTIES = (
-        NAME, IMAGE, COMMAND, CPU, MEMORY,
-        ENVIRONMENT, WORKDIR, LABELS, IMAGE_PULL_POLICY,
-        RESTART_POLICY, INTERACTIVE, IMAGE_DRIVER, HINTS,
-        HOSTNAME, SECURITY_GROUPS, MOUNTS, NETWORKS, TTY,
-    ) = (
-        'name', 'image', 'command', 'cpu', 'memory',
-        'environment', 'workdir', 'labels', 'image_pull_policy',
-        'restart_policy', 'interactive', 'image_driver', 'hints',
-        'hostname', 'security_groups', 'mounts', 'networks', 'tty',
-    )
-
-    _NETWORK_KEYS = (
-        NETWORK_UUID, NETWORK_ID, NETWORK_FIXED_IP, NETWORK_PORT,
-        NETWORK_SUBNET, NETWORK_PORT_EXTRA, NETWORK_FLOATING_IP,
-        ALLOCATE_NETWORK, NIC_TAG,
-    ) = (
-        'uuid', 'network', 'fixed_ip', 'port',
-        'subnet', 'port_extra_properties', 'floating_ip',
-        'allocate_network', 'tag',
-    )
-
-    _IFACE_MANAGED_KEYS = (NETWORK_PORT, NETWORK_ID,
-                           NETWORK_FIXED_IP, NETWORK_SUBNET)
-
-    _MOUNT_KEYS = (
-        VOLUME_ID, MOUNT_PATH, VOLUME_SIZE
-    ) = (
-        'volume_id', 'mount_path', 'volume_size',
-    )
-
-    ATTRIBUTES = (
-        NAME, ADDRESSES
-    ) = (
-        'name', 'addresses'
-    )
-
-    properties_schema = {
-        NAME: properties.Schema(
-            properties.Schema.STRING,
-            _('Name of the container.'),
-            update_allowed=True
-        ),
-        IMAGE: properties.Schema(
-            properties.Schema.STRING,
-            _('Name or ID of the image.'),
-            required=True
-        ),
-        COMMAND: properties.Schema(
-            properties.Schema.STRING,
-            _('Send command to the container.'),
-        ),
-        CPU: properties.Schema(
-            properties.Schema.NUMBER,
-            _('The number of virtual cpus.'),
-            update_allowed=True
-        ),
-        MEMORY: properties.Schema(
-            properties.Schema.INTEGER,
-            _('The container memory size in MiB.'),
-            update_allowed=True
-        ),
-        ENVIRONMENT: properties.Schema(
-            properties.Schema.MAP,
-            _('The environment variables.'),
-        ),
-        WORKDIR: properties.Schema(
-            properties.Schema.STRING,
-            _('The working directory for commands to run in.'),
-        ),
-        LABELS: properties.Schema(
-            properties.Schema.MAP,
-            _('Adds a map of labels to a container. '
-              'May be used multiple times.'),
-        ),
-        IMAGE_PULL_POLICY: properties.Schema(
-            properties.Schema.STRING,
-            _('The policy which determines if the image should '
-              'be pulled prior to starting the container.'),
-            constraints=[
-                constraints.AllowedValues(['ifnotpresent', 'always',
-                                           'never']),
-            ]
-        ),
-        RESTART_POLICY: properties.Schema(
-            properties.Schema.STRING,
-            _('Restart policy to apply when a container exits. Possible '
-              'values are "no", "on-failure[:max-retry]", "always", and '
-              '"unless-stopped".'),
-        ),
-        INTERACTIVE: properties.Schema(
-            properties.Schema.BOOLEAN,
-            _('Keep STDIN open even if not attached.'),
-        ),
-        TTY: properties.Schema(
-            properties.Schema.BOOLEAN,
-            _('Whether the container allocates a TTY for itself.'),
-            support_status=support.SupportStatus(version='14.0.0'),
-        ),
-        IMAGE_DRIVER: properties.Schema(
-            properties.Schema.STRING,
-            _('The image driver to use to pull container image.'),
-            constraints=[
-                constraints.AllowedValues(['docker', 'glance']),
-            ]
-        ),
-        HINTS: properties.Schema(
-            properties.Schema.MAP,
-            _('Arbitrary key-value pairs for scheduler to select host.'),
-            support_status=support.SupportStatus(version='10.0.0'),
-        ),
-        HOSTNAME: properties.Schema(
-            properties.Schema.STRING,
-            _('The hostname of the container.'),
-            support_status=support.SupportStatus(version='10.0.0'),
-        ),
-        SECURITY_GROUPS: properties.Schema(
-            properties.Schema.LIST,
-            _('List of security group names or IDs.'),
-            support_status=support.SupportStatus(version='10.0.0'),
-            default=[]
-        ),
-        MOUNTS: properties.Schema(
-            properties.Schema.LIST,
-            _('A list of volumes mounted inside the container.'),
-            schema=properties.Schema(
-                properties.Schema.MAP,
-                schema={
-                    VOLUME_ID: properties.Schema(
-                        properties.Schema.STRING,
-                        _('The ID or name of the cinder volume mount to '
-                          'the container.'),
-                        constraints=[
-                            constraints.CustomConstraint('cinder.volume')
-                        ]
-                    ),
-                    VOLUME_SIZE: properties.Schema(
-                        properties.Schema.INTEGER,
-                        _('The size of the cinder volume to create.'),
-                    ),
-                    MOUNT_PATH: properties.Schema(
-                        properties.Schema.STRING,
-                        _('The filesystem path inside the container.'),
-                        required=True,
-                    ),
-                },
-            )
-        ),
-        NETWORKS: properties.Schema(
-            properties.Schema.LIST,
-            _('An ordered list of nics to be added to this server, with '
-              'information about connected networks, fixed ips, port etc.'),
-            support_status=support.SupportStatus(version='11.0.0'),
-            schema=properties.Schema(
-                properties.Schema.MAP,
-                schema={
-                    NETWORK_ID: properties.Schema(
-                        properties.Schema.STRING,
-                        _('Name or ID of network to create a port on.'),
-                        constraints=[
-                            constraints.CustomConstraint('neutron.network')
-                        ]
-                    ),
-                    NETWORK_FIXED_IP: properties.Schema(
-                        properties.Schema.STRING,
-                        _('Fixed IP address to specify for the port '
-                          'created on the requested network.'),
-                        constraints=[
-                            constraints.CustomConstraint('ip_addr')
-                        ]
-                    ),
-                    NETWORK_PORT: properties.Schema(
-                        properties.Schema.STRING,
-                        _('ID of an existing port to associate with this '
-                          'container.'),
-                        constraints=[
-                            constraints.CustomConstraint('neutron.port')
-                        ]
-                    ),
-                },
-            ),
-            update_allowed=True,
-        ),
-    }
-
-    attributes_schema = {
-        NAME: attributes.Schema(
-            _('Name of the container.'),
-            type=attributes.Schema.STRING
-        ),
-        ADDRESSES: attributes.Schema(
-            _('A dict of all network addresses with corresponding port_id. '
-              'Each network will have two keys in dict, they are network '
-              'name and network id. '
-              'The port ID may be obtained through the following expression: '
-              '"{get_attr: [<container>, addresses, <network name_or_id>, 0, '
-              'port]}".'),
-            type=attributes.Schema.MAP
-        ),
-    }
-
-    default_client_name = 'zun'
-
-    entity = 'containers'
-
-    def translation_rules(self, props):
-        rules = [
-            translation.TranslationRule(
-                props,
-                translation.TranslationRule.RESOLVE,
-                translation_path=[self.NETWORKS, self.NETWORK_ID],
-                client_plugin=self.client_plugin('neutron'),
-                finder='find_resourceid_by_name_or_id',
-                entity='network'),
-            translation.TranslationRule(
-                props,
-                translation.TranslationRule.RESOLVE,
-                translation_path=[self.NETWORKS, self.NETWORK_PORT],
-                client_plugin=self.client_plugin('neutron'),
-                finder='find_resourceid_by_name_or_id',
-                entity='port')]
-        return rules
-
-    def validate(self):
-        super(Container, self).validate()
-
-        policy = self.properties[self.RESTART_POLICY]
-        if policy and not self._parse_restart_policy(policy):
-            msg = _('restart_policy "%s" is invalid. Valid values are '
-                    '"no", "on-failure[:max-retry]", "always", and '
-                    '"unless-stopped".') % policy
-            raise exception.StackValidationFailed(message=msg)
-
-        mounts = self.properties[self.MOUNTS] or []
-        for mount in mounts:
-            self._validate_mount(mount)
-
-        networks = self.properties[self.NETWORKS] or []
-        for network in networks:
-            self._validate_network(network)
-
-    def _validate_mount(self, mount):
-        volume_id = mount.get(self.VOLUME_ID)
-        volume_size = mount.get(self.VOLUME_SIZE)
-
-        if volume_id is None and volume_size is None:
-            msg = _('One of the properties "%(id)s" or "%(size)s" '
-                    'should be set for the specified mount of '
-                    'container "%(container)s".'
-                    '') % dict(id=self.VOLUME_ID,
-                               size=self.VOLUME_SIZE,
-                               container=self.name)
-            raise exception.StackValidationFailed(message=msg)
-
-        # Don't allow specify volume_id and volume_size at the same time
-        if volume_id and volume_size:
-            raise exception.ResourcePropertyConflict(
-                "/".join([self.NETWORKS, self.VOLUME_ID]),
-                "/".join([self.NETWORKS, self.VOLUME_SIZE]))
-
-    def _validate_network(self, network):
-        net_id = network.get(self.NETWORK_ID)
-        port = network.get(self.NETWORK_PORT)
-        fixed_ip = network.get(self.NETWORK_FIXED_IP)
-
-        if net_id is None and port is None:
-            raise exception.PropertyUnspecifiedError(
-                self.NETWORK_ID, self.NETWORK_PORT)
-
-        # Don't allow specify ip and port at the same time
-        if fixed_ip and port is not None:
-            raise exception.ResourcePropertyConflict(
-                ".".join([self.NETWORKS, self.NETWORK_FIXED_IP]),
-                ".".join([self.NETWORKS, self.NETWORK_PORT]))
-
-    def handle_create(self):
-        args = dict((k, v) for k, v in self.properties.items()
-                    if v is not None)
-        policy = args.pop(self.RESTART_POLICY, None)
-        if policy:
-            args[self.RESTART_POLICY] = self._parse_restart_policy(policy)
-        mounts = args.pop(self.MOUNTS, None)
-        if mounts:
-            args[self.MOUNTS] = self._build_mounts(mounts)
-        networks = args.pop(self.NETWORKS, None)
-        if networks:
-            args['nets'] = self._build_nets(networks)
-        command = args.pop(self.COMMAND, None)
-        if command:
-            args['command'] = shlex.split(command)
-
-        if self.TTY in args:
-            container = self.client(
-                version=self.client_plugin().V1_36).containers.run(**args)
-        else:
-            container = self.client().containers.run(**args)
-        self.resource_id_set(container.uuid)
-        return container.uuid
-
-    def _parse_restart_policy(self, policy):
-        restart_policy = None
-        if ":" in policy:
-            policy, count = policy.split(":")
-            if policy in ['on-failure']:
-                restart_policy = {"Name": policy,
-                                  "MaximumRetryCount": count or '0'}
-        else:
-            if policy in ['always', 'unless-stopped', 'on-failure', 'no']:
-                restart_policy = {"Name": policy, "MaximumRetryCount": '0'}
-
-        return restart_policy
-
-    def _build_mounts(self, mounts):
-        mnts = []
-        for mount in mounts:
-            mnt_info = {'destination': mount[self.MOUNT_PATH]}
-            if mount.get(self.VOLUME_ID):
-                mnt_info['source'] = mount[self.VOLUME_ID]
-            if mount.get(self.VOLUME_SIZE):
-                mnt_info['size'] = mount[self.VOLUME_SIZE]
-            mnts.append(mnt_info)
-        return mnts
-
-    def _build_nets(self, networks):
-        nics = self._build_nics(networks)
-        for nic in nics:
-            net_id = nic.pop('net-id', None)
-            if net_id:
-                nic[self.NETWORK_ID] = net_id
-            port_id = nic.pop('port-id', None)
-            if port_id:
-                nic[self.NETWORK_PORT] = port_id
-
-        return nics
-
-    def check_create_complete(self, id):
-        container = self.client().containers.get(id)
-        if container.status in ('Creating', 'Created'):
-            return False
-        elif container.status == 'Running':
-            return True
-        elif container.status == 'Stopped':
-            if container.interactive:
-                msg = (_("Error in creating container '%(name)s' - "
-                         "interactive mode was enabled but the container "
-                         "has stopped running") % {'name': self.name})
-                raise exception.ResourceInError(
-                    status_reason=msg, resource_status=container.status)
-            return True
-        elif container.status == 'Error':
-            msg = (_("Error in creating container '%(name)s' - %(reason)s")
-                   % {'name': self.name, 'reason': container.status_reason})
-            raise exception.ResourceInError(status_reason=msg,
-                                            resource_status=container.status)
-        else:
-            msg = (_("Unknown status Container '%(name)s' - %(reason)s")
-                   % {'name': self.name, 'reason': container.status_reason})
-            raise exception.ResourceUnknownStatus(status_reason=msg,
-                                                  resource_status=container
-                                                  .status)
-
-    def handle_update(self, json_snippet, tmpl_diff, prop_diff):
-        updaters = []
-        container = None
-
-        after_props = json_snippet.properties(self.properties_schema,
-                                              self.context)
-        if self.NETWORKS in prop_diff:
-            prop_diff.pop(self.NETWORKS)
-            container = self.client().containers.get(self.resource_id)
-            updaters.extend(self._update_networks(container, after_props))
-
-        self.client_plugin().update_container(self.resource_id, **prop_diff)
-
-        return updaters
-
-    def _update_networks(self, container, after_props):
-        updaters = []
-        new_networks = after_props[self.NETWORKS]
-        old_networks = self.properties[self.NETWORKS]
-        security_groups = after_props[self.SECURITY_GROUPS]
-
-        interfaces = self.client(version=self.client_plugin().V1_18).\
-            containers.network_list(self.resource_id)
-        remove_ports, add_nets = self.calculate_networks(
-            old_networks, new_networks, interfaces, security_groups)
-
-        for port in remove_ports:
-            updaters.append(
-                progress.ContainerUpdateProgress(
-                    self.resource_id, 'network_detach',
-                    handler_extra={'args': (port,)},
-                    checker_extra={'args': (port,)})
-            )
-
-        for args in add_nets:
-            updaters.append(
-                progress.ContainerUpdateProgress(
-                    self.resource_id, 'network_attach',
-                    handler_extra={'kwargs': args},
-                    checker_extra={'args': (args['port_id'],)})
-            )
-
-        return updaters
-
-    def check_update_complete(self, updaters):
-        """Push all updaters to completion in list order."""
-        for prg in updaters:
-            if not prg.called:
-                handler = getattr(self.client_plugin(), prg.handler)
-                prg.called = handler(*prg.handler_args,
-                                     **prg.handler_kwargs)
-                return False
-            if not prg.complete:
-                check_complete = getattr(self.client_plugin(), prg.checker)
-                prg.complete = check_complete(*prg.checker_args,
-                                              **prg.checker_kwargs)
-                break
-        status = all(prg.complete for prg in updaters)
-        return status
-
-    def handle_delete(self):
-        if not self.resource_id:
-            return
-        try:
-            self.client().containers.delete(self.resource_id, stop=True)
-            return self.resource_id
-        except Exception as exc:
-            self.client_plugin().ignore_not_found(exc)
-
-    def check_delete_complete(self, id):
-        if not id:
-            return True
-        try:
-            self.client().containers.get(id)
-        except Exception as exc:
-            self.client_plugin().ignore_not_found(exc)
-            return True
-        return False
-
-    def _resolve_attribute(self, name):
-        if self.resource_id is None:
-            return
-        try:
-            container = self.client().containers.get(self.resource_id)
-        except Exception as exc:
-            self.client_plugin().ignore_not_found(exc)
-            return ''
-
-        if name == self.ADDRESSES:
-            return self._extend_addresses(container)
-
-        return getattr(container, name, '')
-
-    def _extend_addresses(self, container):
-        """Method adds network name to list of addresses.
-
-        This method is used only for resolving attributes.
-        """
-        nets = self.neutron().list_networks()['networks']
-        id_name_mapping_on_network = {net['id']: net['name']
-                                      for net in nets}
-        addresses = copy.deepcopy(container.addresses)
-        for net_uuid in container.addresses or {}:
-            addr_list = addresses[net_uuid]
-            net_name = id_name_mapping_on_network.get(net_uuid)
-            if not net_name:
-                continue
-
-            addresses.setdefault(net_name, [])
-            addresses[net_name] += addr_list
-
-        return addresses
-
-
-def resource_mapping():
-    return {
-        'OS::Zun::Container': Container
-    }
--- a/setup.cfg
+++ b/setup.cfg
@@ -87,7 +87,6 @@
 	trove = heat.engine.clients.os.trove:TroveClientPlugin
 	vitrage = heat.engine.clients.os.vitrage:VitrageClientPlugin
 	zaqar = heat.engine.clients.os.zaqar:ZaqarClientPlugin
-	zun = heat.engine.clients.os.zun:ZunClientPlugin
 heat.constraints = 
 	cron_expression = heat.engine.constraint.common_constraints:CRONExpressionConstraint
 	dns_domain = heat.engine.constraint.common_constraints:DNSDomainConstraint
